mlog/human/parser.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package human
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/mattermost/mattermost-server/mlog"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type LogEntry struct {
+	Time    time.Time
+	Level   string
+	Message string
+	Caller  string
+	Fields  []mlog.Field
+}
+
+func ParseLogMessage(msg string) LogEntry {
+	result, err := parseLogMessage(msg)
+	if err != nil {
+		// If failed to parse, just output a LogEntry where all fields are blank, but Message is the original string
+		var result2 LogEntry
+		result2.Message = msg
+		return result2
+	}
+	return result
+}
+
+func parseLogMessage(msg string) (result LogEntry, err error) {
+
+	dec := json.NewDecoder(strings.NewReader(msg))
```

Comment:
I'm a little surprised that zap (what mlog is based on) doesn't provide a way to parse messages into a struct but I guess that's because the log fields are so dynamic that it wouldn't really make sense

Reply:
Yeah I could only find logic for encoding the messages, but not for decoding. Same with logrus.

Also part of the reason this parser is fairly complex is because I wanted to preserve the order the fields were initially output in. Simply parsing the object into a map resulted in non-deterministic output. When I hooked the results up to logrus, I noticed they sort the fields alphabetically.